evm/src/cpu/kernel/interpreter.rs

```diff
@@ -1047,17 +703,15 @@ impl<'a> Interpreter<'a> {
         let new_program_counter =
             u256_to_usize(handler_addr).map_err(|_| anyhow!("The program counter is too large"))?;
 
-        let syscall_info = U256::from(self.generation_state.registers.program_counter + 1)
-            + U256::from((self.generation_state.registers.is_kernel as usize) << 32)
+        let syscall_info = U256::from(self.generation_state.registers.program_counter)
+            + U256::from((self.is_kernel() as usize) << 32)
             + (U256::from(self.generation_state.registers.gas_used) << 192);
         self.generation_state.registers.program_counter = new_program_counter;
 
-        self.generation_state.registers.is_kernel = true;
-        self.kernel_mode = true;
+        *self.is_kernel_mut() = true;
```

Comment:
I think we can save ourselves from future bugs by having a `self.enter_kernel_mode()` (or whatever name you like) to perform these shenanigans.